### ISSUE TITLE
Asserting `DateTime` values between a range in `const` constructors gives an error only when calling

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#54632]
Write this in dartpad.dev:

```dart
class A{
  const A(this.d) : assert(d >= Duration.zero);
  
  final Duration d;
}

void main () {
  const a = A(Duration(days: 1));
}
```

You'll get this error only on line 8:

```
In constant expressions, operands of this operator must be of type 'num'.
```

I tested it on my local development machine and still got the same error. 

My question is, why can't this error be shown in line 2? I do believe this should appear there since packages that do not have tests would never see this and since this example only requires a single `Duration` as a parameter (I have a package that could have a future case like this), this class should not be able to be constant.

Second question, why can't this be asserted in constant expressions even if `Duration`s are constant classes? Can't this be converted to `Duration.inMicroseconds` to assert internally?

**Comment by lrhn** [comment]
It's specified that way because the parser author at the time wanted to be able to recognize potentially constant expressions, without having types available, so potentially constant expressions are defined entirely by syntax.
Then it's a compile time error to evaluate them as constant if the _values_ do not have the right type.

That was a bad idea.
It means you can write a potentially constant expression in a constant constructor which can never be invoked as `const`.
We should fix it, but it's never been a high priority.

**Comment by FMorschel** [comment]
Nowadays, I'm much more familiar with how things work. I'd like to point out two things here, so it is all clear:

- The message on the issue description is from `const_eval_extension_method` (weird name "extension method" since it is not about this at all)
- The error message occurs on `analyzer` when we instantiate the class, so my request here would be to make this evaluation happen on the constructor declaration, not on the constructor invocation (that is what got me to open the issue)

**Comment by lrhn** [comment]
You can't evaluate the expression at the constructor declaration, because it has no binding for the `d` variable until the constructor is invoked with an argument.

(Or are you only suggesting to *report* the error at the constructor instead of at the invocation point? That would probably be useful, since it'll tell the user where something went wrong.)

You can do static analysis at the declaration might be able to tell you that *no* valid argument can make the expression successfully evaluate as a constant expression.
Or not:
```dart
class C {
  const C(Object? o) : assert(o == Duration.zero);
}
void main() {
  const C(null); // Success!
}
```

And it still won't solve all problems, mainly around `dynamic`.
```dart
class C {
  const C(dynamic v) : assert(v.length > 0);
}
void main() {
  const C("42"); // Success.
  C([42]); // Success.
  // const C([42]); // Compile-time error, can only use `.length` on strings.
}
```

That's probably OK, it's typed as `dynamic` so type errors become runtime errors, which for constants is compile-time errors at the execution point.

**Comment by FMorschel** [comment]
> Or are you only suggesting to report the error at the constructor instead of at the invocation point? That would probably be useful, since it'll tell the user where something went wrong

Yes, that is exactly it. I'm asking to remove the need to have an invocation. Thanks for better explaining my intent.

**Comment by lrhn** [comment]
If there is no invocation, then there is no _error_.
Not if the expressions is a valid potentially constant expression, and if it isn't, there is already an error in the constructor.

Take this example again:
```dart
class C {
  const C(Duration? o) : assert(o == Duration.zero);
}
```

That class is highly suspicious, but it's not a compile-time error.
You can *validly* invoke it as:
* `C(null)`
* `C(Duration.zero)`
* `C(Duration(seconds: 1)`
* `const C(null)`.

You can't invoke it as `const C(Duration.zero)`, but trying to do that is a compile-time error just as much as trying to do `const C("Not a Duration!")`, so it's not like you can start thinking that it's not an error.
Where and how to report that error is a usability choice.

We can't give an error at the constructor by itself, because there isn't any error there, not the way the language is currently specified. There are just _some_ invocations that are not valid during constant evaluation.

For `const A(Duration this.d) : assert(d >= Duration.zero);`, it's a _little_ easier to argue that there is an error, because there is no valid argument to a `const` invocation that won't give an error. Any `Duration` instance will not be valid in `d >= Duration.zero`, and even if `d` was an `int`, it won't be a type-valid `int.>=` invocation. So that constructor can *never* be used with `const`. That suggests that the `const` was a mistake.

It would still be a language change to make that constructor be an error.

A *warning* would be reasonable. But then, a warning would be reasonable for any `const` constructor where we can't guarantee that *any* type-valid argument will make the potentially constant expressions be type-valid constant expressions. It's just a question of whether we dare give that warning, and if not, how far do we dare go.

**Comment by FMorschel** [comment]
Yes, you're right. I'd prefer having a warning on the `const` constructor declaration that the `assert` expression can't be evaluated in a `const` context. 

Because if that is the case, having the `const` there is unnecessary, since it will never be able to work.

For example, if we access members, we get this today:

```dart
class C {
  const C.other(String v) : assert(v.isEmpty);
//                                 ^ invalid_constant
}
```

We could have a similar warning when the evaluation there can't be constant.

**Comment by lrhn** [comment]
Yep, the `v.isEmpty` is not a potentially constant expression, so it's an immediate error.
In comparison `assert(v.length == 0)` *is* a potentially constant expression, which is valid as a constant if `v` is a string value (and not if it's not). It's not an error today if it's type-valid, not until you call it as `const` with a non-`String` value.

A `const C(List<int> v) : assert(v.length == 0);` is also incompatible with being `const`.
A `const C(dynamic v) : assert(v.length == 0);` is only potentially incompatible. You can invoke it successfuly, but not with all arguments allowed by the parameter types.

It's a little tricky (probably not impossible) to detect cases where a potentially constant expression will never be valid for constant evaluation (no argument makes it valid). (At least the allowed types are all sealed, so if it's a `Foo`, we know it won't also be a `String`.)
Can't say if it's easier or harder to detect where it *potentially* won't be valid (some argument may make it invalid, but not all - you only have to find one argument as evidence, but that only matters if the types don't rule out everything, so you won't get as much help from the types.).

(Really should fix the language, but that'd be breaking and it's not at the top of the priorities list. Better warnings would help.)

**Comment by eernstg** [comment]
Note the overlap with https://github.com/dart-lang/language/issues/1296. Consider the original example again:

```dart
class A{
  const A(this.d) : assert(d >= Duration.zero);
  final Duration d;
}

const a = A(Duration(days: 1)); // Compile-time error!
```

The failure during constant expression evaluation occurs because the operator `>=` may execute arbitrary code. That is, the receiver type `Duration` isn't among the ones whose behavior is classified as safe, which is only true for a couple of receiver types which are provided by the platform and don't have unknown non-bottom subtypes. Different operators have different rules about those acceptable receiver types, and `>=` only admits `num`, `int` and `double`. In the cases that are supported, the compiler/analyzer can actually execute the given operator implementation at compile time, because it is known, and it has no side effects.

That isn't true for `Duration`'s operator `>=`, that's just a piece of Dart code. So we can't easily support `>=` on `Duration`, but we can easily check whether the receiver type is known to be `num`, `int`, or `double`.

@lrhn wrote a very nice table showing a number of different type checks that we could introduce during the static analysis of constant constructors, in [this comment](https://github.com/dart-lang/language/issues/1296#issuecomment-1110075607).

We  might also allow receivers of type `dynamic` to execute arbitrary potentially constant expressions, in order to avoid breaking code that actually works (even if it isn't safe). A constructor with parameters of type `dynamic` would of course continue to be checked late, that is, at each invocation of the given constructor during constant expression evaluation.

With that, I tend to think this issue is a duplicate of dart-lang/language#1296.

**Comment by FMorschel** [comment]
Thanks for finding it! Yes, I think this is basically a duplicate of that issue. I think I'll close this in favour of that one then.
